### ISSUE TITLE
Backups created by chef-server-ctl backup command cannot be restored after version 12.9.1-1

### ISSUE THREAD

**Issue by chengepaul** [chef/chef-server#1020]
**Description**

Backups created by chef-server-ctl backup command cannot be restored after version 12.9.1-1.

**Chef Version**
12.11.1-1
12.11.0-1
12.10.0-1

Backups created with the latest versions chef-server-core versions cannot be restored.  Chef-server-ctl restore terminates with error message:  "[ERROR] no implicit conversion of nil into String".

**Platform Version**

Ubuntu 14.04

~$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.4 LTS"

Chef-server-core - standalone version

**Replication Case**

install chef-server-core with version 12.11.1-1
chef-server-ctl reconfigure
chef-server-ctl backup -y
chef-server-ctl restore -c chef-backup.tgz

**Stacktrace**

Expanding tarball: /var/opt/chef-backup/chef-backup-2016-11-11-01-06-52.tgz
Restoring Chef Server from backup
Cleaning up any old files
ok: down: bookshelf: 0s, normally up
ok: down: nginx: 0s, normally up
ok: down: oc_bifrost: 1s, normally up
ok: down: oc_id: 0s, normally up
ok: down: opscode-chef-mover: 1s, normally up
ok: down: opscode-erchef: 0s, normally up
ok: down: opscode-expander: 1s, normally up
ok: down: opscode-solr4: 0s, normally up
ok: down: postgresql: 0s, normally up
ok: down: rabbitmq: 1s, normally up
ok: down: redis_lb: 0s, normally up
bookshelf disabled, not stopping
nginx disabled, not stopping
oc_bifrost disabled, not stopping
oc_id disabled, not stopping
opscode-chef-mover disabled, not stopping
opscode-erchef disabled, not stopping
opscode-expander disabled, not stopping
opscode-solr4 disabled, not stopping
postgresql disabled, not stopping
rabbitmq disabled, not stopping
redis_lb disabled, not stopping
Terminating processes running under application users. This will take a few seconds.
Your config files have been backed up to /root/opscode-cleanse-2016-11-11T01:10.
[ERROR] no implicit conversion of nil into String

**Comment by wduncanfraser** [comment]
From what I  saw when testing 12.10, it looks like the backup command is no longer grabbing the /etc directory, just /var. So when it tries to restore there are tons of files missing.

**Comment by irvingpop** [comment]
AFAIK there's no support for restoring a backup made by `chef-server-ctl backup` to a different server version or topology

**Comment by wduncanfraser** [comment]
Correct, however you cannot restore to the same version when /etc/opscode is not backed up.

**Comment by irvingpop** [comment]
Are you sure that /etc/opscode isn't being backed up?   The code is a bit opaque, but doesn't appear to have changed much since I last looked at it:  https://github.com/chef/chef_backup/blob/master/lib/chef_backup/strategy/backup/tar.rb#L84

**Comment by wduncanfraser** [comment]
That's what I saw on 12.10 a couple weeks ago. I'll validate it again on this end and report back my findings.

**Comment by irvingpop** [comment]
Thanks!   when you do, take a look at the manifest file that gets written out in the tarball for confirmation on what was grabbed.

**Comment by shatil** [comment]
@irvingpop I can confirm what @wduncanfraser said. On Ubuntu 14.04, backed up from, and restoring with, `chef-server-core-12.11.1-1`, there is no `/etc/opscode` present:
```
ubuntu@ip-172-31-33-44:~$ tar tvf chef-backup-2016-12-05-20-19-43.tgz | wc -l
9528
ubuntu@ip-172-31-33-44:~$ tar tvf chef-backup-2016-12-05-20-19-43.tgz | grep etc
drwxr-x--- opscode/opscode                      0 2016-12-05 19:43 var/opt/opscode/upgrades/etc/
-rw-r--r-- opscode/opscode                    631 2014-12-06 01:07 var/opt/opscode/upgrades/etc/config.rb
```

I found a similar issue filed a few days ago at https://github.com/chef/chef_backup/issues/26 with the same `[ERROR] no implicit conversion of nil into String` issue as OP.

**Comment by bdwyertech** [comment]
👍

**Comment by bdwyertech** [comment]
Anyone who needs a quick fix on this, see https://github.com/chef/chef_backup/issues/26#issuecomment-268919926

**Comment by stevendanna** [comment]
@bdwyertech's Fix was integrated in Chef Server 12.12 so this should now be fixed.  Thanks @bdwyertech !
